libraries/chain/host_context.cpp

```diff
@@ -101,6 +101,11 @@ int64_t host_context::execute_sync_call(name call_receiver, uint64_t flags, std:
             return handle_call_failure();
          }
 
+         const code_object* const codeobject = db.find<code_object, by_code_hash>(boost::make_tuple(receiver_account->code_hash, receiver_account->vm_type, receiver_account->vm_version));
```

Comment:
Since you already look it up here. Seems like it would be cleaner to pass `code_object` into `execute`. Then `code_hash`, `vm_type`, and `vm_version` do not have to be passed in.

Reply:
`apply_context` uses the same `execute()` and it does not dip `code_object`.  Even if we pass in `code_object` here, we still need to split it and pass in `code_hash`, `vm_type` and `vm_version` before calling `wasm_interface_private::execute()` to reuse the same code.

Reply:
I think `apply_context` might as well look it up and pass it in. In either case it is needed inside `wasm_interface_private::execute()`.

Reply:
It only looks `code_object` when the code is not cached. We will need two `execute()`s with different signatures to avoid unnecessary lookup for `apply_context`. Can we do this refactor in a separate PR?

Reply:
Good point about `code_object` not always needing to be looked up. Maybe we should cache the `sync_call_supported` in `wasm_cache_entry` and `code_descriptor`. That would mean it would have to validated in `execute()`.

Reply:
I think the current implementation is simpler and only checking one location for truth.

Reply:
This extra lookup that must occur on every action bothered me too. It reminds me of EOSIO/eos#9740 since it's a lookup in to an ordered set so I'm worried it'll actually be a visible performance impact for heavy sync calls. 

> Good point about `code_object` not always needing to be looked up. Maybe we should cache the `sync_call_supported` in `wasm_cache_entry` and `code_descriptor`. That would mean it would have to validated in `execute()`.

crossed my mind as well. But that seemed to be _exactly_ what we were trying to avoid by doing it this way since this way is very simple :confused:

Reply:
`code_descriptor` already has `call_offset` so just need to add to `wasm_cache_entry`. I agree the current implementation is simpler, but we avoid the lookup when using the caches so seems like we should continue to avoid the lookup. 

BTW, what happens if you attempt to call the sync call and it isn't there? Do we even need need the verification? Can we just tell by how it fails?

Reply:
> BTW, what happens if you attempt to call the sync call and it isn't there? Do we even need need the verification? Can we just tell by how it fails?

JIT check parameters on every call, so you get some sort of exception out of it. OC does not do any validation so it'll do.. whatever. For example if the WASM call export takes four parameters instead of 3 I guess we'll leak some register's value in to the VM

Reply:
We already have `call_offset` for oc we can check. Seems a bit fragile to rely on the JIT check, so probably should check it. So add it to the `wasm_cache_entry`?

Reply:
In current VM JIT implementation, it will crash if `sync_call()` is not exported. In OC, we have an assert if `code.call_offset` is 0.